Tests/NIOHTTP1Tests/HTTPServerPipelineHandlerTest.swift

```diff
@@ -1181,4 +1181,73 @@ class HTTPServerPipelineHandlerTest: XCTestCase {
         XCTAssertNoThrow(try self.channel.writeOutbound(HTTPServerResponsePart.body(.byteBuffer(ByteBuffer()))))
         XCTAssertNoThrow(try self.channel.writeOutbound(HTTPServerResponsePart.end(nil)))
     }
+
+    func testSendingHeadTwiceGivesError() throws {
+        self.pipelineHandler.failOnPreconditions = false
+        // Sending a head once is normal
+        XCTAssertNoThrow(try self.channel.writeInbound(HTTPServerRequestPart.head(self.requestHead)))
+        // Sending a head twice is an error
+        XCTAssertThrowsError(try self.channel.writeInbound(HTTPServerRequestPart.head(self.requestHead))) { error in
+            XCTAssertEqual(error as? HTTPServerPipelineHandler.ConnectionStateError, .preconditionViolated(message: "received request head in state requestAndResponseEndPending"))
+        }
+    }
+
+    func testServerRespondToNothing() throws {
+        self.pipelineHandler.failOnPreconditions = false
+
+        // Writing an end whilst in state idle is an error
+        XCTAssertThrowsError(try self.channel.writeOutbound(HTTPServerResponsePart.end(nil))) { error in
+            XCTAssertEqual(error as? HTTPServerPipelineHandler.ConnectionStateError, .preconditionViolated(message: "Unexpectedly received a response in state idle"))
+        }
+
+        // Calling finish surfaces the error again
+        XCTAssertThrowsError(try self.channel.finish()) { error in
+            XCTAssertEqual(error as? HTTPServerPipelineHandler.ConnectionStateError, .preconditionViolated(message: "Unexpectedly received a response in state idle"))
+        }
+    }
+
+    func testServerRequestEndFirstIsError() throws {
+        self.pipelineHandler.failOnPreconditions = false
+        // End sending a request which was never startedq
+        XCTAssertThrowsError(try self.channel.writeInbound(HTTPServerRequestPart.end(nil))) { error in
+            XCTAssertEqual(error as? HTTPServerPipelineHandler.ConnectionStateError, .preconditionViolated(message: "Received second request"))
+        }
+    }
+
+    func testForcefulShutdownWhenViolatedPrecondition() async throws {
```

Comment:
We cannot use `EmbeddedChannel` in an async test, but we don't need to. EmbeddedChannel runs strictly synchronously, so tests that use it are 100% reliable. This test should either pass or fail.